Zero copy

boxerab opened the pull request:
Reducing decode memory allocation in this branch:

Decoder code block buffers are no longer allocated: they simply refer to the compressed tile data. To do this, I had to store the code block data  in segments, and only convert into contiguous buffer before t1 decode.

Next step is to use memory mapped stream to no longer allocate compressed tile data, instead have it refer to virtual memory.

boxerab commented:
I need some help troubleshooting the failed clang build:

Errors were found in dynamic analysis log

Everything else is passing.

mayeut commented:
@boxerab, the error in the clang build is reported only there because it's the only build with dynamic analysis enabled. The last errors in the dynamic analysis section of CDASH are showing leaks (but not up to date with you're latest commit, daily quota on CDASH probably reached).
If you're working under windows with Visual Studio, you might wan't to check [Visual Leak Detector](https://vld.codeplex.com) to find the leak.

boxerab commented:
@malaterre  thanks, cool. Except, I ran with Visual Leak Detector and no leaks were detected.
Well, I ran it when decompressing a particular j2k file. I will try this with one of the failed images.

boxerab commented:
so, yeah, I am a bit confused as to which files I should test for memory leaks. I tried a few images but no leaks were detected.

boxerab commented:
@mayeut thanks for the tip. Not sure which files I should test this on.

mayeut commented:
@boxerab the last logs in CDASH (http://my.cdash.org/viewDynamicAnalysisFile.php?id=3390568) shows the following on image issue391.jp2 from the non regression test suite. Also, if you're testing OpenJPEG built as a DLL, be sure to "instrument" the DLL & the EXE ```If your program contains one or more DLLs that you would also like to check for memory leaks, then also include vld.h in at least one source file from each DLL to be included in leak detection.```.

```
ERROR -> opj_decompress: failed to decode image!

=================================================================
==12661==ERROR: LeakSanitizer: detected memory leaks

Direct leak Direct leak of 649728 byte(s) in 1269 object(s) allocated from:
    #0 0x47ab39 in __interceptor_malloc /home/users/aadgrand/LLVM/releases/ubuntu/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:74
    #1 0x7f054ca4883d in opj_vec_init /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/vector.c:41
    #2 0x7f054ca46551 in opj_tcd_code_block_dec_allocate /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1122
    #3 0x7f054ca413bd in opj_tcd_init_tile /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1010
    #4 0x7f054c9e8335 in opj_j2k_read_tile_header /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:8012
    #5 0x7f054c9fd983 in opj_j2k_decode_tiles /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9558
    #6 0x7f054c9e5fe0 in opj_j2k_exec /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:7281
    #7 0x7f054c9ef837 in opj_j2k_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9786
    #8 0x7f054ca0d0b4 in opj_jp2_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/jp2.c:1487
    #9 0x4940f7 in main /home/travis/build/uclouvain/openjpeg/src/bin/jp2/opj_decompress.c:1357
    #10 0x7f054bae876c (/lib/x86_64-linux-gnu/libc.so.6+0x2176c)

Direct leak Direct leak of 43008 byte(s) in 84 object(s) allocated from:
    #0 0x47ab39 in __interceptor_malloc /home/users/aadgrand/LLVM/releases/ubuntu/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:74
    #1 0x7f054ca4883d in opj_vec_init /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/vector.c:41
    #2 0x7f054ca4658a in opj_tcd_code_block_dec_allocate /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1098
    #3 0x7f054ca413bd in opj_tcd_init_tile /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1010
    #4 0x7f054c9e8335 in opj_j2k_read_tile_header /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:8012
    #5 0x7f054c9fd983 in opj_j2k_decode_tiles /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9558
    #6 0x7f054c9e5fe0 in opj_j2k_exec /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:7281
    #7 0x7f054c9ef837 in opj_j2k_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9786
    #8 0x7f054ca0d0b4 in opj_jp2_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/jp2.c:1487
    #9 0x4940f7 in main /home/travis/build/uclouvain/openjpeg/src/bin/jp2/opj_decompress.c:1357
    #10 0x7f054bae876c (/lib/x86_64-linux-gnu/libc.so.6+0x2176c)

Indirect leak Indirect leak of 3648 byte(s) in 114 object(s) allocated from:
    #0 0x47ab39 in __interceptor_malloc /home/users/aadgrand/LLVM/releases/ubuntu/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:74
    #1 0x7f054ca2c49d in opj_seg_buf_add_segment /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/segmented_stream.c:163
    #2 0x7f054ca2c423 in opj_seg_buf_push_segment /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/segmented_stream.c:195
    #3 0x7f054ca3aabe in opj_t2_read_packet_data /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/t2.c:1218
    #4 0x7f054ca379fe in opj_t2_decode_packet /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/t2.c:549
    #5 0x7f054ca372ec in opj_t2_decode_packets /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/t2.c:417
    #6 0x7f054ca43aaa in opj_tcd_t2_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1546
    #7 0x7f054ca43919 in opj_tcd_decode_tile /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/tcd.c:1286
    #8 0x7f054c9ea988 in opj_j2k_decode_tile /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:8061
    #9 0x7f054c9fdaac in opj_j2k_decode_tiles /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9586
    #10 0x7f054c9e5fe0 in opj_j2k_exec /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:7281
    #11 0x7f054c9ef837 in opj_j2k_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/j2k.c:9786
    #12 0x7f054ca0d0b4 in opj_jp2_decode /home/travis/build/uclouvain/openjpeg/src/lib/openjp2/jp2.c:1487
    #13 0x4940f7 in main /home/travis/build/uclouvain/openjpeg/src/bin/jp2/opj_decompress.c:1357
    #14 0x7f054bae876c (/lib/x86_64-linux-gnu/libc.so.6+0x2176c)

SUMMARY: AddressSanitizer: 696384 byte(s) leaked in 1467 allocation(s).
```

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+	/*don't try to read more bytes than are available */
+	bytes_remaining_in_file = seg_buf->data_len - opj_seg_buf_get_global_offset(seg_buf);
+	if (p_nb_bytes > bytes_remaining_in_file) {
+		printf("Warning: attempt to read past end of segmented buffer\n");
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/openjpeg.c:
```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+	fd = open(fname, m, 0666);
+	if (fd < 0) {
+		if (errno > 0 && strerror(errno) != NULL) {
+			printf("%s: %s", fname, strerror(errno));
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/openjpeg.c:
```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+			printf("%s: %s", fname, strerror(errno));
+		}
+		else {
+			printf("%s: Cannot open", fname);
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+		return p_nb_bytes;
+
+	if (p_nb_bytes + opj_seg_buf_get_global_offset(seg_buf)> (OPJ_OFF_T)seg_buf->data_len) {
+		printf("Warning: attempt to skip past end of segmented buffer\n");
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/vector.c:
```diff
@@ -0,0 +1,90 @@
+	vec->size = 0;
+	vec->capacity = VECTOR_INITIAL_CAPACITY;
+	vec->data = opj_malloc(sizeof(void*) * vec->capacity);
+}
```
`opj_malloc` is not checked. No error returned if `NULL`.

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+		return;
+	/*  we allow the offset to move to one location beyond end of buffer segment*/
+	if (buf->offset + off > (OPJ_OFF_T)buf->len) {
+		printf("Warning: attempt to increment buffer offset out of bounds\n");
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/vector.c:
```diff
@@ -0,0 +1,90 @@
+
+void* opj_vec_get(opj_vec_t *vec, OPJ_INT32 index) {
+	if (index >= vec->size || index < 0) {
+		printf("Index %d out of bounds for vec of size %d\n", index, vec->size);
```
There should be no printf in the library code. Please use the existing message mechanism if necessary.

mayeut commented on src/lib/openjp2/vector.c:
```diff
@@ -0,0 +1,90 @@
+static void opj_vec_double_capacity_if_full(opj_vec_t *vec) {
+	if (vec->size >= vec->capacity) {
+		vec->capacity *= 2;
+		vec->data = opj_realloc(vec->data, sizeof(void*) * vec->capacity);
```
`opj_realloc` is not checked. No error returned if `NULL` (& current `vec->data` overriden to `NULL` but not freed in this case).

mayeut commented on src/lib/openjp2/openjpeg.c:
```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+
+	p_source_buffer = (opj_buffer_info_t*)opj_malloc(sizeof(opj_buffer_info_t));
+	if (!p_source_buffer)
+		return NULL;
```
`l_stream` is leaked here.

mayeut commented on src/lib/openjp2/openjpeg.c:
```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+		return NULL;
+
+	buffer_info = (opj_buffer_info_t*)opj_malloc(sizeof(opj_buffer_info_t));
+	memset(buffer_info, 0, sizeof(opj_buffer_info_t));
```
`opj_malloc` return value not checked. Might crash on the next line.

mayeut commented on src/lib/openjp2/openjpeg.c:
```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+	mapped_view = opj_map(fd, buffer_info->len);
+	if (mapped_view == NULL) {
+		opj_memory_map_cleanup(buffer_info);
+		return NULL;
```
`l_stream` is leaked here.

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+		return OPJ_FALSE;
+
+	if (!opj_seg_buf_add_segment(seg_buf, buf, len))
+		return OPJ_FALSE;
```
`buf` is leaked here.

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+
+void opj_seg_buf_incr_cur_seg_offset(opj_seg_buf_t* seg_buf, OPJ_OFF_T offset) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
```
Shouldn't this be an `assert` ?

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+										OPJ_BYTE** ptr,
+										OPJ_SIZE_T chunk_len) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
```
Shouldn't this be an `assert` ?

mayeut commented on src/lib/openjp2/segmented_stream.c:
```diff
@@ -0,0 +1,335 @@
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return OPJ_FALSE;
+	cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
```
given the definition of `opj_vec_get`, `cur_seg` could be `NULL`. An assert would be nice here (or maybe in `opj_vec_get` for bounds checking).

boxerab commented:
Thanks a lot for the code review. I think I've addressed most of these issues.
Unfortunately, still have the memory leak, and Visual Leak Detector is not reporting it.

boxerab commented:
@mayeut plugged the leak! Thanks again for your help.

boxerab commented:
Closing this PR. Turns out that decode performance (at least on windows) is a lot worse on the branch. My guess is the windows memory mapped file implementation. 

So, will need to do a lot more experimentation before this could be merged in.
